src/main/java/pazuzu/service/FeatureService.java

```diff
@@ -0,0 +1,120 @@
+        if (null != existing) {
+            throw new ServiceException("duplicate", "Feature with name " + name + " already exists");
+        }
+
+        final Set<Feature> dependencies = loadFeatures(dependencyNames);
+
+        final Feature newFeature = new Feature();
+        newFeature.setName(name);
+        newFeature.setDockerData(null == dockerData ? "" : dockerData);
+        newFeature.setDependencies(dependencies);
+        featureRepository.save(newFeature);
+        return converter.apply(newFeature);
+    }
+
+    @Transactional(rollbackFor = ServiceException.class)
+    public <T> T updateFeature(String name, String newName, String dockerData, List<String> dependencyNames, Function<Feature, T> converter) throws ServiceException {
+        final Feature existing = loadExistingFeature(name);
+        if (null != newName && !newName.equals(existing.getName())) {
+            final Feature newExisting = featureRepository.findByName(newName);
+            if (null != newExisting) {
+                throw new ServiceException("duplicate", "Feature with name " + newName + " already exists");
+            }
+            existing.setName(newName);
+        }
+        if (null != dockerData) {
+            existing.setDockerData(dockerData);
+        }
+        if (null != dependencyNames) {
+            final Set<Feature> dependencies = loadFeatures(dependencyNames);
+            final List<Feature> recursive = dependencies.stream()
+                    .filter(f -> f.containsDependencyRecursively(existing)).collect(Collectors.toList());
+            if (!recursive.isEmpty()) {
+                throw new ServiceException("recursive", "Recursive dependencies found: " + recursive.stream().map(Feature::getName).collect(Collectors.joining(", ")));
+            }
+            existing.setDependencies(dependencies);
+        }
+        featureRepository.save(existing);
+        return converter.apply(existing);
+    }
+
+    @Transactional
+    public <T> T getFeature(String featureName, Function<Feature, T> converter) throws ServiceException {
+        return converter.apply(loadExistingFeature(featureName));
+    }
+
+    @Transactional(rollbackFor = ServiceException.class)
+    public void deleteFeature(String featureName) throws ServiceException {
+        final Feature feature = featureRepository.findByName(featureName);
+        if (null == feature) {
+            return; // idempotent call
+        }
+        final List<Feature> referencing = featureRepository.findByDependenciesContaining(feature);
+        if (!referencing.isEmpty()) {
+            throw new ServiceException("references", "Can't delete feature " + feature.getName() +
+                    ", references found: " + referencing.stream().map(Feature::getName).collect(Collectors.joining(", ")));
+        }
+        featureRepository.delete(feature);
+    }
+
+    Set<Feature> loadFeatures(List<String> dependencyNames) throws ServiceException {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/Pazuzu/pazuzu-registry/file/2245276191/issues/source?bid=3131965&fileBranchId=3180030#l95)